File: system-tests/tia-client-tia-mode-http/src/main/java/testframework/CustomTestFramework.java
package testframework;

import com.teamscale.report.testwise.model.ETestExecutionResult;
import com.teamscale.report.testwise.model.TestInfo;
import com.teamscale.tia.client.AgentHttpRequestFailedException;
import com.teamscale.tia.client.RunningTest;
import com.teamscale.tia.client.TestRun;
import com.teamscale.tia.client.TiaAgent;
import okhttp3.HttpUrl;
import systemundertest.SystemUnderTest;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/** Simulates a custom test framework that doesn't rely on JUnit. */
public class CustomTestFramework {

	private final Map<String, Runnable> allTests = new HashMap<>();
	public final List<TestInfo> testInfos = new ArrayList<>();
	private final int agentPort;

	public CustomTestFramework(int agentPort) {
		this.agentPort = agentPort;
		allTests.put("testFoo", () -> {
			if (new SystemUnderTest().foo() != 5) {
				throw new AssertionError("Incorrect return value for foo");
			}
		});
		allTests.put("testBar", () -> {
			if (new SystemUnderTest().bar() != 7) {
				throw new AssertionError("Incorrect return value for bar");
			}
		});
	}

	/** Talks to the TIA agent and runs all impacted tests. Also uploads a coverage report at the end. */
	public void runTestsWithTia() throws AgentHttpRequestFailedException {
		TiaAgent agent = new TiaAgent(false, HttpUrl.get("http://localhost:" + agentPort));
		TestRun testRun = agent.startTestRunWithoutTestSelection();

		for (String uniformPath : allTests.keySet()) {
			Runnable runnable = allTests.get(uniformPath);
			RunningTest runningTest = testRun.startTest(uniformPath);

			TestInfo testInfo;
			try {
				runnable.run();
				testInfo = runningTest.endTestAndRetrieveCoverage(
						new TestRun.TestResultWithMessage(ETestExecutionResult.PASSED, ""));
			} catch (Throwable t) {
				testInfo = runningTest.endTestAndRetrieveCoverage(
						new TestRun.TestResultWithMessage(ETestExecutionResult.FAILURE, t.getMessage()));
			}
			testInfos.add(testInfo);
		}
	}

}

File: system-tests/tia-client-tia-mode-http/src/main/kotlin/testframework/CustomTestFramework.kt
package testframework

import com.teamscale.report.testwise.model.ETestExecutionResult
import com.teamscale.report.testwise.model.TestInfo
import com.teamscale.tia.client.AgentHttpRequestFailedException
import com.teamscale.tia.client.TestRun.TestResultWithMessage
import com.teamscale.tia.client.TiaAgent
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrl
import systemundertest.SystemUnderTest

/** Simulates a custom test framework that doesn't rely on JUnit.  */
class CustomTestFramework(private val agentPort: Int) {
	private val allTests = mutableMapOf<String, () -> Unit>()
	val testInfos = mutableListOf<TestInfo>()

	init {
		allTests["testFoo"] = {
			if (SystemUnderTest().foo() != 5) {
				throw AssertionError("Incorrect return value for foo")
			}
		}
		allTests["testBar"] = {
			if (SystemUnderTest().bar() != 7) {
				throw AssertionError("Incorrect return value for bar")
			}
		}
	}

	/** Talks to the TIA agent and runs all impacted tests. Also uploads a coverage report at the end.  */
	@Throws(AgentHttpRequestFailedException::class)
	fun runTestsWithTia() {
		val agent = TiaAgent(false, "http://localhost:$agentPort".toHttpUrl())
		val testRun = agent.startTestRunWithoutTestSelection()

		allTests.keys.forEach { uniformPath ->
			val runnable = allTests[uniformPath]
			val runningTest = testRun.startTest(uniformPath)

			var testInfo: TestInfo
			try {
				runnable?.invoke()
				testInfo = runningTest.endTestAndRetrieveCoverage(
					TestResultWithMessage(ETestExecutionResult.PASSED, "")
				)
			} catch (t: Throwable) {
				testInfo = runningTest.endTestAndRetrieveCoverage(
					TestResultWithMessage(ETestExecutionResult.FAILURE, t.message)
				)
			}
			testInfos.add(testInfo)
		}
	}
}
